src/front/glsl/mod.rs

```diff
@@ -8,140 +8,69 @@ use glsl::{
     syntax::*,
 };
 use parser::{Token, TokenMetadata};
-use std::fmt;
 
 mod helpers;
 mod parser;
 
-#[derive(Debug)]
+#[derive(Debug, thiserror::Error)]
 pub enum ErrorKind {
+    #[error("Unexpected token:\nexpected: {}\ngot: {}", expected.iter().map(|t| t.type_to_string()).collect::<Vec<_>>().join(" |"), got.token)]
```

Comment:
is it idiomatic to use newlines in error strings?

Reply:
Not really I think, but I copied it verbatim from the previous Display trait. I haven't managed to trigger that particular error even when I tried, so I don't know how useful it is currently.